### ISSUE TITLE
isnan() woes

### ISSUE THREAD

**Issue by kwmsmith** [blaze/blaze#1286]
This works:

```
In [43]: df = pd.DataFrame({'a': [1, 0, 0, 1], 'b': [np.nan, np.nan, 0, 0]})

In [44]: x = Data(df, name='x')

In [45]: x
Out[45]:
   a   b
0  1 NaN
1  0 NaN
2  0   0
3  1   0

In [46]: x[(x['a'] == 0) | (x['b'].isnan())]
Out[46]:
   a   b
0  1 NaN
1  0 NaN
2  0   0
```
But this doesn't:

```
In [47]: x.isnan()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-47-9208624ad019> in <module>()
----> 1 x.isnan()

/Users/ksmith/work/blaze/blaze-repo/blaze/expr/expressions.pyc in __getattr__(self, key)
    169             pass
    170         try:
--> 171             result = object.__getattribute__(self, key)
    172         except AttributeError:
    173             fields = dict(zip(map(valid_identifier, self.fields),

AttributeError: 'InteractiveSymbol' object has no attribute 'isnan'
```

And neither does this:

```
In [48]: x['a'].isnan()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-48-641efc28480c> in <module>()
----> 1 x['a'].isnan()

/Users/ksmith/work/blaze/blaze-repo/blaze/expr/expressions.pyc in __getattr__(self, key)
    169             pass
    170         try:
--> 171             result = object.__getattribute__(self, key)
    172         except AttributeError:
    173             fields = dict(zip(map(valid_identifier, self.fields),

AttributeError: 'Field' object has no attribute 'isnan'
```

**Comment by ywang007** [comment]
I Also noticed a case where  isnan is recognized but did not give right answer. 

```python
(Pdb)  type(st.onotional_net)         
<class 'blaze.expr.expressions.Field'>
(Pdb) st.onotional.dshape
dshape("var * ?float64")
(Pdb) st.onotional_net
    onotional_net
0             NaN
1             NaN
(Pdb) st.onotional_net.isnan()
   onotional_net
0          False
1          False

```

**Comment by llllllllll** [comment]
`x['a'].isnan` is not supported because the shape of `x['a']` is `int64`. It would always be false so it can be optimized out at expression time. This is the same for `notnull` (I don't like the lack of symmetry). We could consider optimizing this expression into some 

It would be cool to put these null checking operators on tabular shaped exprs though, something like:

```python
bz.merge(**{k: ds[k].isnan() for k in ds.fields})
```

**Comment by llllllllll** [comment]
@ywang007 what backend is that running with?

**Comment by ywang007** [comment]
Kdb.

**Comment by llllllllll** [comment]
Is that a custom backend?

**Comment by ywang007** [comment]
@llllllllll : the example I gave above is kdb backend; however, I just tried with sqlite and seeming something a bit different:

```
>>> from blaze import Data
>>> d = Data('sqlite:///./myfile.db::test')
>>> d
          A
0       NaN
1 -0.227388
2  1.531978
3       NaN
>>> d.dshape
dshape("var * {A: ?float64}")
>>> type(d.A)
<class 'blaze.expr.expressions.Field'>
>>> d.A.isnan()
      A
0  None
1  None
2  None
3  None

```

**Comment by llllllllll** [comment]
hmm, I tried in postgres:

```python
In [1]: ds = bz.Data('postgresql://localhost/test::floattest')

In [2]: ds.a
Out[2]: 
      a
0  None
1  None
2  None
3     1
4     2
5     3
6   NaN
7   NaN
8   NaN

In [3]: ds.a.isnan()
Out[3]: 
       a
0   None
1   None
2   None
3  False
4  False
5  False
6   True
7   True
8   True
```

which looks correct; however, I agree that what you got in sqlite looks incorrect.


About the kdb issue: I don't believe that the kdb backend is part of blaze by default. Do you have some extension installed?

**Comment by cpcloud** [comment]
`isnan` doesn't make sense as an operation on anything except `float64`, `float32` and `?float64`, `?float32` columns

it's behaving as expected by saying "hey this operation isn't available". i think this makes sense

all bets are kind of off with sqlite here because there isn't really such a thing as a nan. the things that you might expect to return a nan such as `0.0 / 0.0` return a `NULL` so sqlite's `isnan` impl should probably just be a call to `isnull` instead

**Comment by llllllllll** [comment]
Maybe we should raise an exception on sql's isnan if the dialect is sqlite? I think that making isnan->isnull would be confusing for columns of type `?float`. For example, the table I showed with a nan, null, and value.

**Comment by cpcloud** [comment]
yeah that's a better idea, and probably in the error message suggest using `isnull()`

**Comment by cpcloud** [comment]
i also don't htink we have an `isnull` :)

**Comment by llllllllll** [comment]
I think we chose to just do a `notnull`. Having the `isnan` and `notnull` be flipped is nice because it makes people think about the difference. I could see an argument for making an `isnull` though. This seems like it might be a part of the docs that need to be expanded on to talk about the difference between a null and a nan especially because some backends use them interchangably.

**Comment by kwmsmith** [comment]
> i also don't htink we have an isnull :)

Yes, I think implementing `isnull()` and `notnull()` with the same semantics as Pandas' would address the underlying issue here.

Specifically, it would be able to be applied to non-float columns, or entire tables, without raising an exception, just like in Pandas.

My vote would be to deprecate and eventually remove `isnan()` once `isnull` and `notnull` are in place.

**Comment by llllllllll** [comment]
`isnull` and `isnan` are very different operations. `NaN` is a valid element of type `float`, where null is not a valid element of type `float`. The missing value for options types should not be mixed with nan.

**Comment by cpcloud** [comment]
> Yes, I think implementing `isnull()` and `notnull()` with the same semantics as Pandas' would address the underlying issue here.

Implementing `isnull()` and `notnull()` with the exact same semantics as pandas on all backends defeats the purpose of having two distinct functions `isnan` and `notnull`. Pandas treats `NaN`s as `NULL`s because NumPy's type system cannot handle missing values.

For example,

With this table definition I can certainly have `NaN` values, and no `NULL`s
```sql
create table nans (
    bad_calcs double precision not null
);

insert into nans values ('nan'::float8), (1.0);
insert into nans values (NULL);  -- raises an error
```

And with this one, I can have both kinds of values

```sql
create table nulls_and_nans (
    bad_calcs_et_al double precision
);

insert into nulls_and_nans values ('nan'::float8), (NULL);
```

**Comment by kwmsmith** [comment]
> isnull and isnan are very different operations. NaN is a valid element of type float, where null is not a valid element of type float. The missing value for options types should not be mixed with nan.

Yes, I understand, thanks for emphasizing the distinction here.

Because we have backends where NaN and null are representable and distinct, and other python-centric backends where they are conflated, would it make sense to have both `isnull()` and `isnan()` in Blaze?  For SQL backends, they would respect the distinction, while for Pandas backends, `isnull == isnan` is one possible implementation.

One downside--and I'd expect this to arise in other scenarios as well--is that the same expression could give different results when running on different backends with the same logical data that has both nans and nulls present.

**Comment by kwmsmith** [comment]
> Implementing isnull() and notnull() with the exact same semantics as pandas on all backends defeats the purpose of having two distinct functions isnan and notnull. Pandas treats NaNs as NULLs because NumPy's type system cannot handle missing values.

Yes, that's why I was in favor of deprecating `isnan` once `isnull` is in place.

But @llllllllll 's comment and your explanation make me realize the NaN / NULL distinction is important to maintain in Blaze, so I retract my naive suggestion :)

What about implementing both `isnan` and `isnull` (and their inverses), and for Pandas / NumPy backends, `isnan` is mapped to `isnull`?  Or does that open a pandora's box?

Having backend-dependent behavior for the same logical dataset has to have come up in other situations in blaze--has it?  If so, what's the guideline / principle in those scenarios?
